python/lsst/daf/butler/_query_results.py

```diff
@@ -0,0 +1,749 @@
+            the full query and then throwing away the result rows after
+            counting them.  If `False`, this is an error, as the user would
+            usually be better off executing the query first to fetch its rows
+            into a new query (or passing ``exact=False``).  Ignored if
+            ``exact=False``.
+
+        Returns
+        -------
+        count : `int`
+            The number of rows the query would return, or an upper bound if
+            ``exact=False``.
+
+        Notes
+        -----
+        This counts the number of rows returned, not the number of unique rows
+        returned, so even with ``exact=True`` it may provide only an upper
+        bound on the number of *deduplicated* result rows.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def any(self, *, execute: bool = True, exact: bool = True) -> bool:
+        """Test whether this query returns any results.
+
+        Parameters
+        ----------
+        execute : `bool`, optional
+            If `True`, execute at least a ``LIMIT 1`` query if it cannot be
+            determined prior to execution that the query would return no rows.
+        exact : `bool`, optional
+            If `True`, run the full query and perform post-query filtering if
+            needed, until at least one result row is found.  If `False`, the
+            returned result does not account for post-query filtering, and
+            hence may be `True` even when all result rows would be filtered
+            out.
+
+        Returns
+        -------
+        any : `bool`
+            `True` if the query would (or might, depending on arguments) yield
+            result rows.  `False` if it definitely would not.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def explain_no_results(self, execute: bool = True) -> Iterable[str]:
+        """Return human-readable messages that may help explain why the query
+        yields no results.
+
+        Parameters
+        ----------
+        execute : `bool`, optional
+            If `True` (default) execute simplified versions (e.g. ``LIMIT 1``)
+            of aspects of the tree to more precisely determine where rows were
+            filtered out.
+
+        Returns
+        -------
+        messages : `~collections.abc.Iterable` [ `str` ]
+            String messages that describe reasons the query might not yield any
+            results.
+        """
+        raise NotImplementedError()
+
+
+class ParentDatasetQueryResults(DatasetQueryResults):
+    """An object that represents results from a query for datasets with a
+    single parent `DatasetType`.
+    """
+
+    @property
+    @abstractmethod
+    def components(self) -> Sequence[str | None]:
+        """The components of the parent dataset type included in these results
+        (`~collections.abc.Sequence` [ `str` or `None` ]).
+        """
+        raise NotImplementedError()
```

Comment:
I think we can drop this from the new API.  If its equivalent in the old API wasn't deprecated, I think it should have been, as RFC-879 included deprecating all support for components in query APIs.

Reply:
`Registry.queryDatasets`'s `component` parameter is documented as deprecated, I guess that makes that attribute deprecated too. I'll drop it in the new API and I can explicitly document it as deprecated in old API.